pkg/jwks/jwks.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package jwks manages downloading and updating the keys from a JWKS source
+// for keys.
+package jwks
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/exposure-notifications-server/internal/database"
+	hadb "github.com/google/exposure-notifications-server/internal/verification/database"
+	"github.com/google/exposure-notifications-server/internal/verification/model"
+	"github.com/google/exposure-notifications-server/pkg/logging"
+	"github.com/rakutentech/jwk-go/jwk"
+)
+
+type Config struct {
+	Database database.Config
+}
+
+func (c *Config) DatabaseConfig() *database.Config {
+	return &c.Database
+}
+
+func (c *Config) JWKSManagerConfig() *Config {
+	return c
+}
+
+// Manager handles updating all HealthAuthorities if they've specified a JWKS
+// URI.
+type Manager struct {
+	ctxt context.Context
+	db   *database.DB
+}
+
+// getKeys reads the keys for a single HealthAuthority from its jwks server.
+func (mgr Manager) getKeys(ha *model.HealthAuthority) ([]byte, error) {
+	if len(ha.JwksURI) == 0 {
+		return nil, nil
+	}
+
+	reqCtxt, done := context.WithTimeout(mgr.ctxt, 5*time.Second)
+	defer done()
+	req, err := http.NewRequestWithContext(reqCtxt, "GET", ha.JwksURI, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating connection: %w", err)
+	}
+
+	var resp *http.Response
+	resp, err = http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("reading connection: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("resp (%v) != %v", resp.StatusCode, http.StatusOK)
+	}
+
+	var bytes []byte
+	bytes, err = ioutil.ReadAll(resp.Body)
```

Comment:
I think it'd be more efficient to stream unmarshall directly into the []jwt.JWT instead of having the []byte as an interim.

Reply:
I did it this way because it's far easier to test. As written the `getKeys` and `parseKeys` are separable. (Additionally, it's _likely_ [not guaranteed] there isn't much inefficiency...) If you feel passionately enough, I'll end up combining these two functions.

Reply:
N is probably sufficiently small that it's not worth it. Let's do this for now.